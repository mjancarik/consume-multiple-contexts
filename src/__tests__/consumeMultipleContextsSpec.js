import React from 'react';
import { mount } from 'enzyme';
import {
  createMultipleContexts,
  createNamedContext,
  withMultipleContextsFactory
} from '../main';

const ThemeContext = React.createContext('light');
const UserContext = React.createContext({ name: 'profile' });

class Profile extends React.Component {
  render() {
    return (
      <div>
        {this.props.theme}, {this.props.user && this.props.user.name}
      </div>
    );
  }
}

class Other extends React.Component {
  render() {
    return (
      <span>
        {this.props.theme}, {this.props.user && this.props.user.name}
      </span>
    );
  }
}

function pure(props) {
  return (
      <i>
      {props.theme}, {props.user && props.user.name}
     </i>
 );
}

class App extends React.Component {
  render() {
    const { user, theme, Component } = this.props;

    return (
      <ThemeContext.Provider value={theme}>
        <UserContext.Provider value={user}>
          <Component />
        </UserContext.Provider>
      </ThemeContext.Provider>
    );
  }
}

describe('consume-multiple-contexts', () => {
  describe('createNamedContext method', () => {
    it('should return right structure', () => {
      expect(createNamedContext('theme', ThemeContext)).toMatchSnapshot();
    });
  });

  describe('createMultipleContexts method', () => {
    let withContext = null;

    beforeAll(() => {
      withContext = createMultipleContexts(
        createNamedContext('theme', ThemeContext),
        createNamedContext('user', UserContext)
      );
    });

    it('should create multiple contexts factory', () => {
      function Web() {
        return withContext(({ theme, user }) => (
          <Profile theme={theme} user={user} />
        ));
      }

      const wrapper = mount(<App theme="dark" Component={Web} />);

      expect(wrapper).toMatchSnapshot();
      wrapper.setProps({ theme: 'red', user: { name: 'anonymous' } });
      expect(wrapper).toMatchSnapshot();
    });

    it('should create multiple contexts factory 2', () => {
      function Web() {
        return withContext(({ theme, user }) => (
          <Other theme={theme} user={user} />
        ));
      }

      expect(mount(<Web />)).toMatchSnapshot();
    });
  });

  describe('withMultipleContextsFactory HOC', () => {
    it('should pass context to props for React.Component', () => {
      const Component = withMultipleContextsFactory(
        createNamedContext('theme', ThemeContext),
        createNamedContext('user', UserContext)
      )(Profile);

      const wrapper = mount(
        <App theme="dark" user={{ name: 'anonymous' }} Component={Component} />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should pass context to props for pure function', () => {
        const Component = withMultipleContextsFactory(
          createNamedContext('theme', ThemeContext),
          createNamedContext('user', UserContext)
      )(pure);

        const wrapper = mount(
          <App theme="dark" user={{ name: 'anonymous' }} Component={Component} />
        );

        expect(wrapper).toMatchSnapshot();
    });
  });
});
