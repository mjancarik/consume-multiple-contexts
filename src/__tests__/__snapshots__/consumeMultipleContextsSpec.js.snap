// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consume-multiple-contexts createMultipleContexts method should create multiple contexts factory 1`] = `
<App
  Component={[Function]}
  theme="dark"
>
  <Web>
    <Profile
      theme="dark"
    >
      <div>
        dark
        , 
      </div>
    </Profile>
  </Web>
</App>
`;

exports[`consume-multiple-contexts createMultipleContexts method should create multiple contexts factory 2`] = `
<App
  Component={[Function]}
  theme="red"
  user={
    Object {
      "name": "anonymous",
    }
  }
>
  <Web>
    <Profile
      theme="red"
      user={
        Object {
          "name": "anonymous",
        }
      }
    >
      <div>
        red
        , 
        anonymous
      </div>
    </Profile>
  </Web>
</App>
`;

exports[`consume-multiple-contexts createMultipleContexts method should create multiple contexts factory 2 1`] = `
<Web>
  <Other
    theme="light"
    user={
      Object {
        "name": "profile",
      }
    }
  >
    <span>
      light
      , 
      profile
    </span>
  </Other>
</Web>
`;

exports[`consume-multiple-contexts createNamedContext method should return right structure 1`] = `
Object {
  "Consumer": Object {
    "$$typeof": Symbol(react.context),
    "Consumer": [Circular],
    "Provider": Object {
      "$$typeof": Symbol(react.provider),
      "_context": [Circular],
    },
    "_calculateChangedBits": null,
    "_changedBits": 0,
    "_changedBits2": 0,
    "_currentRenderer": null,
    "_currentRenderer2": null,
    "_currentValue": "light",
    "_currentValue2": "light",
    "_defaultValue": "light",
  },
  "name": "theme",
}
`;

exports[`consume-multiple-contexts withMultipleContextsFactory HOC should pass context to props for React.Component 1`] = `
<App
  Component={[Function]}
  theme="dark"
  user={
    Object {
      "name": "anonymous",
    }
  }
>
  <ComponentWithMultipleContexts>
    <Profile
      theme="dark"
      user={
        Object {
          "name": "anonymous",
        }
      }
    >
      <div>
        dark
        , 
        anonymous
      </div>
    </Profile>
  </ComponentWithMultipleContexts>
</App>
`;

exports[`consume-multiple-contexts withMultipleContextsFactory HOC should pass context to props for pure function 1`] = `
<App
  Component={[Function]}
  theme="dark"
  user={
    Object {
      "name": "anonymous",
    }
  }
>
  <ComponentWithMultipleContexts>
    <pure
      theme="dark"
      user={
        Object {
          "name": "anonymous",
        }
      }
    >
      <i>
        dark
        , 
        anonymous
      </i>
    </pure>
  </ComponentWithMultipleContexts>
</App>
`;
